orttraining/orttraining/core/optimizer/memory_alleviation.cc

```diff
@@ -0,0 +1,754 @@
+
+          LOGS(logger, VERBOSE) << "Node " << parent_node.Name() << "(" << parent_node.OpType() << ")'s "
+                                << parent_node_output_index
+                                << "th output [" << parent_node.OutputDefs()[parent_node_output_index]->Name()
+                                << "] is added in recompute search list  ";
+
+          q.push_back(next_p);
+        }
+      }
+    }
+    // After handle all entry node outputs, we set the flag to false.
+    is_entry_node_output_arg = false;
+  }
+
+  // If input args are not found in bw, but op count exceed MAXIMUM_RECOMPUTE_NODE_COUNT, skip recompute.
+  if (!q.empty() || early_stop) {
+    LOGS(logger, VERBOSE) << "Fail to find a solution for recompute: current node count is " << nodes.size()
+                          << ", queue size: " << q.size() << ", early stop: " << early_stop;
+    nodes.clear();
+  } else {
+    // Re-order the nodes in topological order.
+    std::sort(nodes.begin(), nodes.end(),
+              [&node_index_to_its_order_in_topological_sort_map](const Node*& lhs, const Node*& rhs) {
+                return node_index_to_its_order_in_topological_sort_map.at(lhs->Index()) <
+                       node_index_to_its_order_in_topological_sort_map.at(rhs->Index());
+              });
+  }
+  return Status::OK();
+}
+
+bool MemoryAlleviation::IsNodeRecomputable(const Node& node,
+                                           const ActivationUsedMap& fw_op_output_arg_used_map,
+                                           const InlinedHashMap<NodeIndex, size_t>&
+                                               node_index_to_its_order_in_topological_sort_map,
+                                           const InlinedHashMap<const Node*, InlinedVector<size_t>>&
+                                               candidate_output_args_map,
+                                           AlleviationSubGraphStores& subgraph_stores,
+                                           const logging::Logger& logger,
+                                           bool compromise_stashed_activation,
+                                           bool& can_compromise_stashed_activation) const {
+  InlinedVector<const Node*> nodes_in_topological_order;
+  ORT_ENFORCE(SelectRecomputeSubgraph(node, candidate_output_args_map.at(&node),
+                                      fw_op_output_arg_used_map,
+                                      node_index_to_its_order_in_topological_sort_map,
+                                      nodes_in_topological_order, logger,
+                                      compromise_stashed_activation,
+                                      can_compromise_stashed_activation)
+                  .IsOK());
+  if (nodes_in_topological_order.size() == 0) {
+    return false;
+  }
+
+  std::string subgraph_str_representation, log_info;
+  NodesInTypoOrderToString(nodes_in_topological_order, subgraph_str_representation, log_info);
```

Comment:
nit: typo
NodesInTypoOrderToString --> NodesInTopoOrderToString

Reply:
Done, thank you!